KEP-961: Bump Statefulset MaxUnavailable to beta (built on top of #4474) (#5228)
kubernetes/enhancements: keps/sig-apps/961-maxunavailable-for-statefulset/README.md

conscious choice that more than one pod going down at the same time during rolling update, would not
cause issues with their Stateful Application which have per pod state and identity. Other Stateful
Applications which cannot tolerate more than one pod going down, will resort to the current behavior of one pod at a time Rolling Updates.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

StatefulSet RollingUpdate strategy will contain an additional parameter called `maxUnavailable` to
control how many Pods will be brought down at a time, during Rolling Update.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

None.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> 

Reviewer: We need specific details of the proposal, something like:
```
This document proposes to add new field to StatefulSet's RollingUpdate configuration 
settings (`.spec.updateStrategy.rollingUpdate`) called `maxUnavailable`, which will 
allow setting the maximum number of pods that can be unavailable during the update.
This new field can be an absolute, positive number (eg. `5`) or a percentage of desired 
pods (ex: `10%`). When the field is not specified, it will default to 1 to maintain 
previous behavior of rolling one pod at a time.
```

### User Stories

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

As a User of Kubernetes, I should be able to update my StatefulSet, more than one Pod at a time, in a
RollingUpdate manner, if my Stateful app can tolerate more than one pod being down, thus allowing my
update to finish much faster.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

> With `maxUnavailable` feature enabled, we'll bring down more than one pod at a time, if your application can't tolerate

Reviewer: ```suggestion
With `maxUnavailable` field set, we'll bring down more than one pod at a time, if your application can't tolerate
```

this behavior, you should absolutely leave the field unset or use 1, which will match default behavior.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

We are proposing a new field called `maxUnavailable` whose default value will be 1. In this mode, StatefulSet will behave exactly like its current behavior.

> Its possible we introduce a bug in the implementation. The mitigation currently is that is disabled by default in Alpha phase for people to try out and give

Reviewer: ```suggestion
It's possible we introduce a bug in the implementation. The mitigation currently is that this feature is disabled by default in Alpha phase for people to try out and give
```

feedback.

> In Beta phase when its enabled by default, people will only see issues or bugs when `maxUnavailable` is set to something greater than 1. Since people have

Reviewer: ```suggestion
In Beta phase when its enabled by default, users will only see issues or bugs when `maxUnavailable` is set to something greater than 1. Since people have
```

tried this feature in Alpha, we would have time to fix issues.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> ### Implementation Details

Reviewer: Nit: while editing this document, can you drop the json and protobuf tags from the API below leaving ONLY the type information.

        // Defaults to 1.
        // +optional
        MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,2,opt,name=maxUnavailable"`

	...
}
```

- By Default, if maxUnavailable is not specified, its value will be assumed to be 1 and StatefulSets
  will follow their old behavior. This will also help while upgrading from a release which doesnt support maxUnavailable to a release which supports this field.
- If maxUnavailable is specified, it cannot be greater than total number of replicas.
- If maxUnavailable is specified and partition is also specified, MaxUnavailable cannot be greater than `replicas-partition`
- If a partition is specified, maxUnavailable will only apply to all the pods which are staged by the
  partition. Which means all Pods with an ordinal that is greater than or equal to the partition will be
  updated when the StatefulSet’s .spec.template is updated. Lets say total replicas is 5 and partition is set to 2 and maxUnavailable is set to 2. If the image is changed in this scenario, following

>   are the possible behavior choices we have:-

Reviewer: ```suggestion
  are the possible behavior choices we have:
```

>   1.  Pods with ordinal 4 and 3 will start Terminating at the same time(because of maxUnavailable). Once they are both running and available, pods with ordinal 2 will start Terminating. Pods with ordinal 0 and 1

Reviewer: While editing, can you fix this sentence above:
from:
```
This will also help while upgrading from a release which doesnt support maxUnavailable to a release which supports this field
```
to
```
This will also help while upgrading from a release which doesn't support maxUnavailable to a release which supports this field
```

Reviewer: One more from:
```
Lets say total replicas is 5 and partition is set to 2 and maxUnavailable is set to 2
```
to 
```
Let's say total replicas is 5 and partition is set to 2 and maxUnavailable is set to 2
```

>   1.  Pods with ordinal 4 and 3 will start Terminating at the same time(because of maxUnavailable). Once they are both running and available, pods with ordinal 2 will start Terminating. Pods with ordinal 0 and 1

Reviewer: ```suggestion
  1.  Pods with ordinal 4 and 3 will start Terminating at the same time (because of maxUnavailable). Once they are both running and available, pods with ordinal 2 will start Terminating. Pods with ordinal 0 and 1
```

      will remain untouched due the partition. In this choice, the number of pods terminating is not always
      maxUnavailable, but sometimes less than that. For e.g. if pod with ordinal 3 is running and available but 4 is not, we still wait for 4 to be running and available before moving on to 2. This implementation avoids
      out of order Terminations of pods.

>   2.  Pods with ordinal 4 and 3 will start Terminating at the same time(because of maxUnavailable). When any of 4 or 3 are running and available, pods with ordinal 2 will start Terminating. This could violate

Reviewer: ```suggestion
  2.  Pods with ordinal 4 and 3 will start Terminating at the same time (because of maxUnavailable). When any of 4 or 3 are running and available, pods with ordinal 2 will start Terminating. This could violate
```

      ordering guarantees, since if 3 is running and available, then both 4 and 2 are terminating at the same

>       time out of order. If 4 is running and available, then both 3 and 2 are Terminating at the same time and no ordering guarantees are violated. This implementation, guarantees, that always there are maxUnavailable number of Pods Terminating except the last batch.

Reviewer: ```suggestion
      time out of order. If 4 is running and available, then both 3 and 2 are Terminating at the same time and no ordering guarantees are violated. This implementation guarantees, that always there are maxUnavailable number of Pods Terminating except the last batch.
```

>   3.  Pod with ordinal 4 and 3 will start Terminating at the same time(because of maxUnavailable). When 4 is running and available, 2 will start Terminating. At this time both 2 and 3 are terminating. If 3 is

Reviewer: ```suggestion
  3.  Pod with ordinal 4 and 3 will start Terminating at the same time (because of maxUnavailable). When 4 is running and available, 2 will start Terminating. At this time both 2 and 3 are terminating. If 3 is
```

>       running and available before 4, 2 wont start Terminating to preserve ordering semantics. So at this time,

Reviewer: ```suggestion
      running and available before 4, 2 won't start Terminating to preserve ordering semantics. So at this time,
```

      only 1 is unavailable although we requested 2.

>   4.  (Implemented approach) Introduce a field in Rolling Update, which decides whether we want maxUnavailable with ordering or without ordering guarantees. Depending on what the user wants, this Choice can either choose behavior 1 or 3 if ordering guarantees are needed or choose behavior 2 if they dont care. To simplify this further

Reviewer: ```suggestion
  4.  (Implemented approach) Introduce a field in Rolling Update, which decides whether we want maxUnavailable with ordering or without ordering guarantees. Depending on what the user wants, this Choice can either choose behavior 1 or 3 if ordering guarantees are needed or choose behavior 2 if they don't care. To simplify this further
```

>       PodManagementPolicy today supports `OrderedReady` or `Parallel`. The `Parallel` mode only supports scale up and tear down of StatefulSets and currently doesnt apply to Rolling Updates. So instead of coming up

Reviewer: ```suggestion
      PodManagementPolicy today supports `OrderedReady` or `Parallel`. The `Parallel` mode only supports scale up and tear down of StatefulSets and currently doesn't apply to Rolling Updates. So instead of coming up
```

difference in the ordinals which are Terminating out of Order, cannot be more than maxUnavailable.

Choice 2 always offers maxUnavailable number of Pods in Terminating state. This can sometime lead to
pods terminating out of order. This will always lead to the fastest rollouts. The maximum difference in the ordinals which are Terminating out of Order, can be more than maxUnavailable.

Choice 3 always guarantees than no two pods are ever Terminating out of order. It sometimes does that,
at the cost of not being able to Terminate maxUnavailable pods. The implementation for this might be
complicated.

Choice 4 provides a choice to the users and hence takes the guessing out of the picture on what they
will expect. Implementing Choice 4 using PMP would be the easiest.

#### Implementation

The alpha release we are going with Choice 4 with support for both PMP=Parallel and PMP=OrderedReady.
For PMP=Parallel, we will use Choice 2.
For PMP=OrderedReady, the plan for alpha was to go with Choice 3 to ensure we can support ordering guarantees while also
making sure the rolling updates are fast, but for simplicity Choice 1 was implemented instead.

> We are keeping implementation the same for beta release, going the simpler route, but instead of checking for healthy pods in the part of the code that decides what is a `unavailablePod`, we check for `isUnavailable`, fixing https://github.com/kubernetes/kubernetes/issues/112307.

Reviewer 2: Why not use 4+3 (instead of 4+1), to ensure we can support ordering guarantees while also making sure the rolling updates are fast

Author: Depends on whether we prefer to have this feature but having a relatively complex algorithm to determine the target. 

It can get quite complex with maxUnavailable > 2. Let's say replica = 5, maxUnavailable = 3. For the first round, index 4,3,2 are restarted. We now need to keep track of what pods are currently unavailable, say `unavailablePodList`, and `If unavailablePods < maxUnavailable` we need to perform extra checks.  `if unavailablePods == 2`, then we need to check that index 4 isn't in `unavailablePodList` in order to proceed with 1. `if unavailablePods == 1`, we need to check that index 4 and index 3 are not in `unavailablePodList` to proceed with 1 and 0. If index 3 is in `unavailablePodList` but 4 isn't, then we can only restart 1. If both are in, or only 4 is in `unavailablePodList`, nothing can proceed. 

This is a brute force way and I'm sure there is a simpler way to do it, but this example does highlight that it can get harder to track how many to restart. 

Open to discussions on whether or not it is worth pursuing.

> https://github.com/kubernetes/kubernetes/blob/eb8f3f194fed16484162aebdaab69168e02f8cb4/pkg/controller/statefulset/stateful_set_control.go#L740

Author: fixed

...
```

> New proposed implementation:

Author: changed to PR

- maxUnavailable greater than 1 with partition and staged pods greater than maxUnavailable
- maxUnavailable greater than 1 with partition and maxUnavailable greater than replicas

New testcases being added:

- Feature enablement/disablement test

Coverage:

- `pkg/apis/apps/v1`: `2023-05-26` - `71.7%`
- `pkg/apis/apps/v1beta2`: `2023-05-26` - `76.7%`
- `pkg/apis/apps/validation`: `2023-05-26` - `92.3%`
- `pkg/controller/statefulset`: `2023-05-26` - `85.7%`

> - `pkg/registry/apps/statefulset`: `2023-05-26` - `63.9%`

Reviewer: Can you please update the coverage results?

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- `test/integration/statefulset`: we'll add a test to verify that the number of pods brought down each
  time is less or equal to configured maxUnavailable and podManagementPolicy settings are honored.

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- `test/e2e/apps/statefulset.go`: 
  - test that rolling updates are working correctly for both PodManagementPolicy types when the MaxUnavailable is used. 
  - include a test that fails currently but passes when https://github.com/kubernetes/kubernetes/issues/112307 is fixed, with a 

>     StatefulSet setting `minReadySeconds` and `updateStrategy.rollingUpdate.maxUnavailable` and checking for a correct rollout specially when scaling down during a rollout.

Reviewer: The lack of integration and e2e for this feature might be a blocker for beta promotion. With [recent changes to beta requirements](https://github.com/kubernetes/enhancements/pull/5242) this will be a problem. Unless you'll be able to land tests prior to merging this KEP for promotion. Given you have time until [June 20th](https://github.com/kubernetes/sig-release/tree/master/releases/release-1.34#summary) that's achievable.

- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

#### Upgrade

- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
  - This is a new field, it will only be enabled after an upgrade. To maintain the
    previous behavior, you can disable the feature gate manually in Beta or leave
    the field unconfigured.
  - This field was set in alpha before upgrading:
    - If the flag was enabled, it continue to be enabled in Beta
    - If the flag was disabled, it will continue to be disbled in Beta
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
  - This feature is enabled by default in Beta so you can configure the maxUnavailable
    to meet your requirements

#### Downgrade

- If we try to set this field, it will be silently dropped.
- If the object with this field already exists, we maintain is as it is, but the controller
  will just ignore this field.

> - downgrade of kube-apiserver to a version with the feature flag was disabled, when the field was previously set - the field gets ignored by controller, and controller behaves as if the field is not set

Reviewer: ```suggestion
- downgrade of kube-apiserver to a version with the feature flag was disabled, when the field was previously set - the field gets ignored by apiserver, and controller behaves as if the field is not set
```

- `pkg/apis/apps/v1beta2`: `2023-05-26` - `76.7%`
- `pkg/apis/apps/validation`: `2023-05-26` - `92.3%`
- `pkg/controller/statefulset`: `2023-05-26` - `85.7%`
- `pkg/registry/apps/statefulset`: `2023-05-26` - `63.9%`

##### Integration tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- `test/integration/statefulset`: we'll add a test to verify that the number of pods brought down each
  time is less or equal to configured maxUnavailable and podManagementPolicy settings are honored.

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> - `test/e2e/apps/statefulset.go`: - test that rolling updates are working correctly for both PodManagementPolicy types when the MaxUnavailable is used. - include a test that fails currently but passes when https://github.com/kubernetes/kubernetes/issues/112307 is fixed, with a

Author: fixed

- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Initial support for maxUnavailable in StatefulSets added. Disabled by default with default value of 1.

#### Beta

- Enabled by default with default value of 1 with upgrade/downgrade tested at least manually.
- Added `statefulset_unavailability_violation` metric in-tree
- It is necessary to update the firstUnhealthyPod calculation to correctly call processCondemned. New tests should cover this and take into consideration that the controller should first wait for the predecessor condemned pods to become available before deleting them and delete the pod with the highest ordinal number
- minReadySeconds and maxUnavailable bugs https://github.com/kubernetes/kubernetes/issues/123911, https://github.com/kubernetes/kubernetes/issues/112307, https://github.com/kubernetes/kubernetes/issues/119234 and https://github.com/kubernetes/kubernetes/issues/123918 should be fixed before promotion of maxUnavailable.
- Additional unit/e2e/integration tests listed in the test plan should be added covering the newly found bugs.

> - Users should be warned that maxUnavailable works differently for each podManagementPolicy (E.g for OrderedReady it is not applied until the StatefulSet had a chance to fully scale up). This can result in slower rollouts. For parallel this can skip ordering. This should be both mentioned in the API doc and website as a requirements for beta graduation.

Author: I can add it as an action item for Beta graduation

> - Users should be warned that maxUnavailable works differently for each podManagementPolicy (E.g for OrderedReady it is not applied until the StatefulSet had a chance to fully scale up). This can result in slower rollouts. For parallel this can skip ordering. This should be both mentioned in the API doc and website as a requirements for beta graduation.

Reviewer: Have you started working on the above? As I mentioned elsewhere, we should not wait for this document update to land, but try to handle the problems that exists right away. Especially the tests will block eventual graduation of this KEP.

#### GA

- No critical bugs reported and no bad feedbacks collected.

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

#### Upgrade

- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
  - This is a new field, it will only be enabled after an upgrade. To maintain the
    previous behavior, you can disable the feature gate manually in Beta or leave

>     the field unconfigured.

Author: Added both cases, PTAL

- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
  - This feature is enabled by default in Beta so you can configure the maxUnavailable
    to meet your requirements

#### Downgrade

- If we try to set this field, it will be silently dropped.
- If the object with this field already exists, we maintain is as it is, but the controller
  will just ignore this field.

> - downgrade of kube-apiserver to a version with the feature flag was disabled, when the field was previously set - it gets cleared, new operations setting the field will be silently dropped

Author: oh yeah, fixed

- downgrade of kube-controller-manager to version with the feature flag disabled, when the field was previously set - the field gets ignored by controller, and controller behaves as if the field is not set
- downgrade of either kube-apiserver or kube-controller-manager to a version with the feature flag was disabled, when the field was not set the behavior doesn't change

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and
  in the kubelet? How does an n-2 kubelet without this feature available behave
  when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

> 

Author: changed feature enabled -> version with feature enabled

###### Are there any tests for feature enablement/disablement?

<!--
The e2e framework does not currently support enabling or disabling feature
gates. However, unit tests in each component dealing with managing data, created
with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.

Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

There are unit tests which make sure the field is correctly dropped
on feature enable and disabled, see [strategy tests](https://github.com/kubernetes/kubernetes/blob/23698d3e9f4f3b9738ba5a6fcefd17894a00624f/pkg/registry/apps/statefulset/strategy_test.go#L391-L417).

> Feature enablement/disablement test will also be added when graduating to Beta as [TestStatefulSetStartOrdinalEnablement](https://github.com/kubernetes/kubernetes/blob/23698d3e9f4f3b9738ba5a6fcefd17894a00624f/pkg/registry/apps/statefulset/strategy_test.go#L473)

Reviewer: This is a hard requirement for beta, as mentioned in other comments. So you need to add that one asap.

  - StatefulSet updates revert to the default Kubernetes behavior, ignoring any `maxUnavailable` settings even if previously specified.

In scenarios 1 and 2, running workloads could be impacted due to the mismatch in feature flag settings. For instance, expected rolling update strategies might not be applied, potentially affecting application availability or update speed.

To mitigate these risks, it's crucial to ensure consistent feature flag settings across all control plane components during rollout and to be prepared to adjust configurations based on observed behavior. Monitoring the progress and outcomes of StatefulSet updates closely during the initial adoption phase is recommended to quickly identify and resolve any issues.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

Administrators should monitor the following metrics to assess the need for a rollback:

- `statefulset_unavailability_violation`: Ths metric reflects the number of times the maxUnavailable condition is violated (i.e. spec.replicas - availableReplicas > maxUnavailable).

> Mutiple violations of maxUnavailable might indicate issues with feature behavior.

Reviewer: ```suggestion
Multiple violations of maxUnavailable might indicate issues with feature behavior.
```

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

If their StatefulSet rollingUpdate section has the field maxUnavailable specified with
a value different than 1. While in alpha and beta, the feature-gate needs to be enabled.

The command bellow should show the maxUnavailable value:

```
kubectl get statefulsets -o yaml | grep maxUnavailable
```

###### How can someone using this feature know that it is working for their instance?

Users can verify the maxUnavailable feature is working for their StatefulSets by observing the behavior of their application during rolling updates. Specifically, they should monitor the update progress of StatefulSets against the maxUnavailable value they have configured. This can be done by tracking the number of pods that are unavailable at any given time during the update process and ensuring it does not exceed the specified maxUnavailable limit. Users can also view the `statefulset_unavailability_violation` metric to see if there have been instances

> where the feature is not working as intended.

Reviewer: You want to follow the [template](https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md#how-can-an-operator-determine-if-the-feature-is-in-use-by-workloads) here and add something like:
```
- [ ] Events
  - Event Reason: 
- [x] API .spec
  - Condition name: 
  - Other field: .spec.updateStrategy.rollingUpdate.maxUnavailable
- [ ] Other (treat as last resort)
  - Details:
 ```

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

Startup latency of schedulable stateful pods should follow the [existing latency SLOs](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos).

> 

Reviewer: How about stating a reasonable SLO for that new `statefulset_unavailability_violation` metric?

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, you can disable the feature-gate manually once it's in Beta, this will affect StatefulSets that are in the middle of a rollout when the feature gets disabled.

> If the feature is disabled in the middle of the rollout, the rollout will continue as if maxUnavailable is 1.

Author: Added

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes, you can disable the feature-gate manually once it's in Beta, this will affect StatefulSets that are in the middle of a rollout when the feature gets disabled.

Author: added

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

The rollout or rollback of the `maxUnavailable` feature for StatefulSets primarily affects how updates are managed, aiming to minimize disruptions. However, several scenarios could lead to potential issues:

Incorrect `maxUnavailable` Configurations: Setting `maxUnavailable` too high or too low could lead to service disruptions or slow update processes, respectively.

Feature Flag Consistency Across Control Plane: Inconsistencies in applying the feature flag MaxUnavailableStatefulSet across the control plane components, especially in HA clusters, can lead to unpredictable behaviors. The impact varies based on the combination of the feature flag state (on/off) for kube-apiserver and kube-controller-manager:

1. Feature Enabled on kube-apiserver and Disabled on kube-controller-manager:

> - The maxUnavailable setting is accepted by the API server but ignored by the controller. This could lead to slower rollouts than expected because the controller doesn't act on the provided `maxUnavailable` values.

Reviewer 2: Nit: add indentation so that it's nested under the bullet point above

Author: fixed

To mitigate these risks, it's crucial to ensure consistent feature flag settings across all control plane components during rollout and to be prepared to adjust configurations based on observed behavior. Monitoring the progress and outcomes of StatefulSet updates closely during the initial adoption phase is recommended to quickly identify and resolve any issues.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

Administrators should monitor the following metrics to assess the need for a rollback:

- `statefulset_unavailability_violation`: Ths metric reflects the number of times the maxUnavailable condition is violated (i.e. spec.replicas - availableReplicas > maxUnavailable).
Mutiple violations of maxUnavailable might indicate issues with feature behavior.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

A manual test was performed, as follows:

> 1. Create a cluster in 1.23.

Reviewer 2: 1.23 is pretty old. Why not use a newer version?

Author: moved to 1.33 instead

2. Upgrade to 1.24.
3. Create StatefulSet A with spec.updateStrategy.rollingUpdate.maxUnavailable set to 3, with 6 replicas
4. Verify a rollout and check if only 3 pods are unavailable at a time ([currently with a bug if podManagementPolicy is set to Parallel](https://github.com/kubernetes/kubernetes/issues/112307))
5. Downgrade to 1.23.
6. Verify that the rollout only has 1 pod unavailable at a time, similar to setting maxUnavailable to 1
7. Create another StatefulSet B not setting maxUnavailable (leaving it nil)
8. Upgrade to 1.24.
9. Verify that the rollout has default behavior of only having one pod unavailable at a time
   Verify that the `maxUnavailable` can be set again to StatefulSet A and test the rollout behavior

TODO:
A manual test will be performed, as follows:

1. Create a cluster in 1.29.

> 2. Upgrade to 1.31.

Reviewer 2: In general we'd upgrade to N+1 (instead of N+2). Why are these 2 versions picked?

3. Create StatefulSet A with spec.updateStrategy.rollingUpdate.maxUnavailable set to 3, with 6 replicas
4. Verify a rollout and check if only 3 pods are unavailable at a time
5. Check if rollout is also fine with podManagementPolicy set to Parallel
6. Downgrade to 1.29.
7. Verify that the rollout only has 1 pod unavailable at a time, similar to setting maxUnavailable to 1 (MaxUnavailableStatefulSet feature gate disabled by default).
8. Create another StatefulSet B not setting maxUnavailable (leaving it nil)
9. Upgrade to 1.31.
10. Verify that the rollout has default behavior of only having one pod unavailable at a time
    Verify that the `maxUnavailable` can be set again to StatefulSet A and test the rollout behavior

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

If their StatefulSet rollingUpdate section has the field maxUnavailable specified with

> a value different than 1. While in alpha and beta, the feature-gate needs to be enabled.

Reviewer 2: You can still set it to 1 when this feature is enabled, right? Perhaps they just need to check the existence of the maxUnavailable field, and that the feature gate is enabled.

Author: true, it can be both. During rollout though you won't see a behavioral change if it is set to 1, but you will see it if it set to anything else. As an example, LWS sets the value even if the feature flag was disabled. The only way we could tell it wasn't enabled by default was by testing it with a value larger than one.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

The RollingUpdate will fail or will not be able to proceed if etcd or API server is unavailable and
hence this feature will also be not be able to be used.

###### What are other known failure modes?

<!--
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

- Incorrect Handling of minReadySeconds During StatefulSet Updates with Parallel Pod Management
  - Detection:

>     - Monitor the status.availableReplicas of the StatefulSet during rolling updates. A significant drop below the expected number of available replicas, considering the maxUnavailable setting, could indicate the issue.

Reviewer 2: Can the new metric be leveraged here?

Author: yes, added as a way to detect

    - Temporarily adjust the podManagementPolicy to OrderedReady as a workaround to ensure minReadySeconds is respected during updates, though this may slow down the rollout process.
    - Setting maxUnavailable back to 1 or disabling the feature is a possibility.
    - Upgrade your cluster to 1.31, where this feature is promoted to beta, and the issue is fixed.
  - Diagnostics:
    - Events should be the first place one should look when trying to diagnose this failure mode.
    - If events are not sufficient, detailed logging could be enabled for the StatefulSet controller to capture the sequence and timing of pod updates during a rollout.
    - Look for patterns or log entries indicating that a new pod update is initiated before the minReadySeconds period has elapsed for the previously updated pod.
  - Testing: e2e tests for this issue will be added as part of this promotion.

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

- 2019-01-01: KEP created.
- 2019-08-30: PR Implemented with tests covered.
- <<[UNRESOLVED bugs found in alpha and blockers to promotion @knelasevero @atiratree @bersalazar @leomichalski]>>

>   Will be able to update this section when I have PRs merged and the dates for that.

Reviewer 2: Link to the issues here

Author: added link to PRs
